<a href="https://colab.research.google.com/github/Nniikkoollaass/data-science-modul-10-HW/blob/main/DS_HW_10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
pip install scikit-surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 9.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3162981 sha256=d38bc1878c84039340dcbbd6589458f878eff6fc0e1d72f9b33aa8b4d4953ea1
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


Завдання

На цьому тижні ми вивчили як працюють рекомендаційні системи. Пропонуємо вам познайомитись з бібліотекою surprise, котра якраз є по суті додатком до знайомої нам бібліотеки scikit-learn для тренування моделей рекомендаційних систем.

Візьміть датасет movielens і побудуйте модель матричної факторизації. У даній бібліотеці він має назву SVD. Підберіть найкращі параметри за допомогою крос-валідації, також поекспериментуйте з іншими алгоритмами розрахунків (SVD++, NMF) і оберіть той, який буде оптимальним.

Підказки як саме побудувати дану модель ви знайдете в документації до даної бібліотеки.

In [2]:
from surprise import Reader, Dataset, SVD, NMF
from surprise.model_selection import cross_validate

In [6]:
# Задаємо структуру даних
reader = Reader(line_format='user item rating timestamp', sep=',', rating_scale=(0.5, 5), skip_lines=1)

# Завантажуємо дані
data = Dataset.load_from_file('ml-latest-small/ratings.csv', reader=reader)

# Вибираємо алгоритми
algorithms = {
    'SVD': SVD(),
    'NMF': NMF()
}

# Запускаємо крос-валідацію для кожного алгоритму
for algo_name, algo in algorithms.items():
    print(f'Running {algo_name}...')
    results = cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
    print(f'{algo_name} results:')
    for metric, values in results.items():
        print(f'{metric}:')
        for value in values:
            print(f'\t{value:.3f}')

Running SVD...
Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8741  0.8676  0.8723  0.8722  0.8764  0.8725  0.0029  
MAE (testset)     0.6715  0.6663  0.6694  0.6743  0.6709  0.6705  0.0026  
Fit time          1.41    1.42    1.43    1.39    2.20    1.57    0.31    
Test time         0.12    0.12    0.20    0.22    0.21    0.18    0.04    
SVD results:
test_rmse:
	0.874
	0.868
	0.872
	0.872
	0.876
test_mae:
	0.672
	0.666
	0.669
	0.674
	0.671
fit_time:
	1.406
	1.423
	1.428
	1.388
	2.196
test_time:
	0.124
	0.123
	0.198
	0.222
	0.210
Running NMF...
Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9228  0.9209  0.9152  0.9195  0.9271  0.9211  0.0039  
MAE (testset)     0.7070  0.7040  0.7009  0.7044  0.7103  0.7053  0.0032  
Fit time          2.48    2.46    2.44    3.19    2.53    2.62    0

Зазвичай вибір між алгоритмами рекомендаційних систем здійснюється на основі показників якості, таких як RMSE (Root Mean Squared Error) та MAE (Mean Absolute Error), де менше значення вказує на кращу точність моделі.

В результатах SVD має середнє значення RMSE приблизно 0.8725, а NMF має приблизно 0.9211. Таким чином, SVD виявляється кращим алгоритмом за RMSE.

Також, якщо порівнювати MAE, SVD має середнє значення приблизно 0.6705, а NMF - приблизно 0.7053. Отже, і за MAE також SVD виявляється кращим.

Отже, на основі обох метрик (RMSE та MAE) можна визначити, що алгоритм SVD демонструє кращі результати порівняно з NMF на цьому наборі даних. Тому, для цієї моделі рекомендаційної системи рекомендується використовувати SVD.